support drag & drop outside of NPE and implement copy & paste

campersau opened the pull request:
fixes https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/issues/383

It is now also possible to copy `PackageFolder`s inside of NPE.

campersau commented on PackageExplorer/PackageViewer.xaml.cs:
```diff
@@ -489,13 +470,109 @@ private void PackageFolderContextMenu_Opened(object sender, RoutedEventArgs e)
+        private IDataObject CreateDataObject(PackagePart packagePart)
+        {
+            var data = new DataObject();
+            data.SetData(PackageFileDataFormat, packagePart.Path);
```
During drag & drop the object does not have to be `Serializable` but for copy & past it is required.
That's why now the `PackageFileDataFormat` is just the path and not the actual `PackagePart`.
Since drag & drop from one NPE instance into another one didn't work before this is not a breaking change.

campersau commented on PackageExplorer/PackageViewer.xaml.cs:
```diff
@@ -489,13 +470,109 @@ private void PackageFolderContextMenu_Opened(object sender, RoutedEventArgs e)
+
+                    var packagePart = RootFolder.GetPackageParts().FirstOrDefault(part => part.Path == packagePartPath);
+
+                    if (packagePart != null)
```
This and other things currently prevent drag & drop / copy & past from one NPE instance into another one.

onovotny commented:
LGTM, lmk when you're done or if you have more changes you need in this PR

campersau commented:
I also added copy & paste menu items. Files can be copied. In folders can be pasted.

![image](https://user-images.githubusercontent.com/4009570/37870000-c13dd68e-2fc3-11e8-9092-97aabdf78b74.png)

![image](https://user-images.githubusercontent.com/4009570/37870002-c4325270-2fc3-11e8-90f7-ce437c0a46cc.png)

It also supports copying from other windows applications like RDP / Outlook via IStream.

campersau commented:
@onovotny ready for merge!

onovotny commented:
Is there a place for the new `Span<T>` or `Memory<T>` in any of the interop code?

campersau commented:
@onovotny Yes, it looks like `Marshal.Copy` can be replaced with `Span`s and `Unsafe.ToPointer` but then we have to compile with `/unsafe`. See https://github.com/campersau/NuGetPackageExplorer/commit/37039787c72056f2e9dc6d743292003001ebb524?w=1

benaadams commented:
> and `Unsafe.ToPointer` but then we have to compile with /unsafe

Would also need to fix the span or array over the `Unsafe.AsPointer` + `Marshal.StructureToPtr` call; else you have a gc hole and the `IntPtr` may not point to the correct location for the `StructureToPtr` call

campersau commented:
@benaadams Do you mean something like this? Also if we need to fix them, we could just do it directly without `Span`s:

Without Span
```csharp
unsafe
{
    fixed (byte* p = bytes)
    {
        Marshal.StructureToPtr(source, (IntPtr)p, false);
    }
}
```

With Span
```csharp
unsafe
{
    fixed (byte* p = bytes)
    {
        MarshalStructure(new ReadOnlySpan<byte>(p, bytes.Length));
    }
}

return bytes;

unsafe void MarshalStructure(ReadOnlySpan<byte> fixedBytes)
{
    var p = Unsafe.AsPointer(ref MemoryMarshal.GetReference(fixedBytes));

    Marshal.StructureToPtr(source, (IntPtr)p, false);
}
````

---

Without Span
```csharp
unsafe
{
    fixed (byte* p = fileGroupDescriptorBytes)
    {
        return MarshalFileNames(p);
    }
}

unsafe string[] MarshalFileNames(byte* fixedBytes)
{
    var fileGroupDescriptor = Marshal.PtrToStructure<FILEGROUPDESCRIPTORW>((IntPtr)fixedBytes);
    var fileNames = new string[fileGroupDescriptor.cItems];

    fixedBytes = fixedBytes + Marshal.SizeOf<FILEGROUPDESCRIPTORW>();

    for (var fileDescriptorIndex = 0; fileDescriptorIndex < fileGroupDescriptor.cItems; fileDescriptorIndex++)
    {
        var fileDescriptor = Marshal.PtrToStructure<FILEDESCRIPTORW>((IntPtr)fixedBytes);
        fileNames[fileDescriptorIndex] = fileDescriptor.cFileName;

        fixedBytes = fixedBytes + Marshal.SizeOf<FILEDESCRIPTORW>();
    }
    return fileNames;
}
```

With Span
```csharp
unsafe
{
    fixed (byte* p = fileGroupDescriptorBytes)
    {
        return MarshalFileNames(new ReadOnlySpan<byte>(p, fileGroupDescriptorBytes.Length));
    }
}

unsafe string[] MarshalFileNames(ReadOnlySpan<byte> fixedBytes)
{
    var p = Unsafe.AsPointer(ref MemoryMarshal.GetReference(fixedBytes));

    var fileGroupDescriptor = Marshal.PtrToStructure<FILEGROUPDESCRIPTORW>((IntPtr)p);
    var fileNames = new string[fileGroupDescriptor.cItems];

    p = Unsafe.Add<byte>(p, Marshal.SizeOf<FILEGROUPDESCRIPTORW>());

    for (var fileDescriptorIndex = 0; fileDescriptorIndex < fileGroupDescriptor.cItems; fileDescriptorIndex++)
    {
        var fileDescriptor = Marshal.PtrToStructure<FILEDESCRIPTORW>((IntPtr)p);
        fileNames[fileDescriptorIndex] = fileDescriptor.cFileName;

        p = Unsafe.Add<byte>(p, Marshal.SizeOf<FILEDESCRIPTORW>());
    }
    return fileNames;
}
```

benaadams commented:
Can just fix the return from `MemoryMarshal.GetReference` and keep it all in the one method:
```csharp
unsafe void MarshalStructure(ReadOnlySpan<byte> bytes)
{
    fixed (byte *p = &MemoryMarshal.GetReference(bytes))
    {
        Marshal.StructureToPtr(source, (IntPtr)p, false);
    }
}
```
The return from `Unsafe.AsPointer` can be moved after the call if its not already fixed so become incorrect. `fixed` is close to zero cost *if* no gc happens during the fixed time.

benaadams commented:
Though if you don't have anything fancy going on in the struct (e.g. `FILEGROUPDESCRIPTORW` doesn't look exciting) could use `Unsafe.As`

`FILEDESCRIPTORW` on the other hand is an exciting struct as it has a `string` in it so you would still need to marshal that

benaadams commented:
Something like?
```csharp
unsafe string[] MarshalFileNames(ReadOnlySpan<byte> bytes)
{
    ref var data = ref MemoryMarshal.GetReference(bytes);
    var fileGroupDescriptor = Unsafe.As<byte, FILEGROUPDESCRIPTORW>(ref data);
    var fileNames = new string[fileGroupDescriptor.cItems];

    fixed (byte* pStart = &Unsafe.Add(ref data, Marshal.SizeOf<FILEGROUPDESCRIPTORW>()))
    {
        var ptr = pStart;

        for (var fileDescriptorIndex = 0; fileDescriptorIndex < fileGroupDescriptor.cItems; fileDescriptorIndex++)
        {
            var fileDescriptor = Marshal.PtrToStructure<FILEDESCRIPTORW>((IntPtr)ptr);
            fileNames[fileDescriptorIndex] = fileDescriptor.cFileName;

            ptr += Marshal.SizeOf<FILEDESCRIPTORW>();
        }
    }

    return fileNames;
}
```

campersau commented:
Thanks @benaadams !
Here is your suggestion https://github.com/campersau/NuGetPackageExplorer/commit/1a3bccddd8790a71f8d8cbdb12f46348775e61d9?diff=split&w=1

Let's see what @onovotny thinks about the 3 alternatives and if this is worth the additional dependencies & unsafe code.

onovotny commented:
TBH, I don't have any strong feeling either way. I'm okay with using `unsafe`, but beyond that, I'm good with whatever makes the most sense for readability and reducing potential errors in the code. I don't have a strong sense of what the best approach is in various circumstances yet, hence the question :)

What do you both think?

benaadams commented:
> Here is your suggestion campersau/NuGetPackageExplorer@1a3bccd?diff=split&w=1

You can drop the outer `unsafe`/`fixed`; also make use of implicit conversion to ReadOnlySpan so
```csharp
var fileGroupDescriptorStream = (MemoryStream)data.GetData(FileGroupDescriptorW);
var fileGroupDescriptorBytes = fileGroupDescriptorStream.ToArray();
 	 
unsafe
{
    fixed (byte* p = fileGroupDescriptorBytes)
    {
        return MarshalFileNames(new ReadOnlySpan<byte>(p, fileGroupDescriptorBytes.Length));
    }
}
```
Could become:
```csharp
var fileGroupDescriptorStream = (MemoryStream)data.GetData(FileGroupDescriptorW);
var fileGroupDescriptorBytes = fileGroupDescriptorStream.ToArray();
return MarshalFileNames(fileGroupDescriptorBytes);
```

benaadams commented:
> What do you both think?

Personally I prefer staying on the managed heap rather than using `Marshal.AllocHGlobal` + `Marshal.FreeHGlobal`; but I have an aversion to manually deallocating memory 😄

onovotny commented:
@benaadams I agree, I like the managed heap much better, in general.

campersau commented:
@benaadams So fixing is only required if we are dealing with pointers and not with refs?

Then we also don't need the local function right?

```csharp
ReadOnlySpan<byte> fileGroupDescriptorBytes = fileGroupDescriptorStream.ToArray();

ref var fileGroupDescriptorPtr = ref MemoryMarshal.GetReference(fileGroupDescriptorBytes);
var fileGroupDescriptor = Unsafe.As<byte, FILEGROUPDESCRIPTORW>(ref fileGroupDescriptorPtr);

var fileNames = new string[fileGroupDescriptor.cItems];
unsafe
{
    fixed (byte* pStart = &Unsafe.Add(ref fileGroupDescriptorPtr, Marshal.SizeOf<FILEGROUPDESCRIPTORW>()))
    {
        var fileDescriptorRowPtr = pStart;
        for (var fileDescriptorIndex = 0; fileDescriptorIndex < fileGroupDescriptor.cItems; fileDescriptorIndex++)
        {
            var fileDescriptor = Marshal.PtrToStructure<FILEDESCRIPTORW>((IntPtr)fileDescriptorRowPtr);
            fileNames[fileDescriptorIndex] = fileDescriptor.cFileName;

            fileDescriptorRowPtr += Marshal.SizeOf<FILEDESCRIPTORW>();
        }
    }
}
return fileNames;
```

benaadams commented:
> So fixing is only required if we are dealing with pointers and not with refs?

Yeah, the GC will track the `ref`s and update them if it moves memory, also is aware they are in use. 

As far as the runtime is concerned pointers and `IntPtr` are just numbers so have no coupling to the actual memory; so you need to either fix the memory, use a pinned GC handle or use native allocation (which is fixed) to stop it potentially moving.

benaadams commented on PackageExplorer/Interop/NativeDragDrop.cs:
```diff
@@ -0,0 +1,315 @@
+
+            public override long Seek(long offset, SeekOrigin origin)
+            {
+                var ptr = Marshal.AllocHGlobal(Marshal.SizeOf<ulong>());
```
Since you are going unsafe... 😉

This could just use stack memory (stack is also fixed memory): 
```csharp
public unsafe override long Seek(long offset, SeekOrigin origin)
{
    long result = 0L;
    var ptr = &result;

    _inner.Seek(offset, (int)origin, (IntPtr)ptr);

    return result;
}
```

benaadams commented on PackageExplorer/Interop/NativeDragDrop.cs:
```diff
@@ -0,0 +1,315 @@
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                var ptr = Marshal.AllocHGlobal(Marshal.SizeOf<ulong>());
```
Could also use stack here instead
```csharp
public unsafe override int Read(byte[] buffer, int offset, int count)
{
    long result = 0L;
    var ptr = &result;

    if (offset > 0)
    {
        buffer = buffer.Skip(offset).ToArray();
    }

    _inner.Read(buffer, count, (IntPtr)ptr);

    return (int)result;
}
```

benaadams commented on PackageExplorer/Interop/NativeDragDrop.cs:
```diff
@@ -0,0 +1,315 @@
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                var ptr = Marshal.AllocHGlobal(Marshal.SizeOf<ulong>());
```
Also could use stack here
```csharp
public unsafe override void Write(byte[] buffer, int offset, int count)
{
    long result = 0L;
    var ptr = &result;

    if (offset > 0)
    {
        buffer = buffer.Skip(offset).ToArray();
    }

    _inner.Write(buffer, count, (IntPtr)ptr);

    var written = (int)result;

    if (written != count)
    {
        Write(buffer, written, count - written);
    }
}
```

campersau commented:
@benaadams Thanks for the clarification!

@onovotny I updated the code to use the managed heap instead. You can merge if you are ok with it.

benaadams commented on PackageExplorer/Interop/NativeDragDrop.cs:
```diff
@@ -204,68 +204,51 @@ public override void Flush()
+                }
 
-                    _inner.Write(buffer, count, ptr);
+                var written = 0;
```
Should this be a `long` rather than `int`? Don't know the api that well

campersau replied:
In the docs I coun't find anything about the type https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.comtypes.istream.write(v=vs.110).aspx

But I thought that if the `count` is an `int` then the result can't be bigger.

benaadams replied:
Yeah docs for Read has **`ULONG`** in bold for the return type https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.comtypes.istream.read(v=vs.110).aspx

Suppose if it works 😄

benaadams replied:
Ah it says refer to `ISequentialStream::Write` which does say its a pointer to a `ulong` https://msdn.microsoft.com/en-us/library/windows/desktop/aa380014(v=vs.85).aspx

However as the other param is also a ulong its likely a C ulong; meaning they are all C# uint/int sizes. Same for the other methods

onovotny commented:
Thanks! merging this in now, will wait a little bit before releasing to the store/choco in case there's stuff that creeps in.
